flink-optimizer/src/main/java/org/apache/flink/optimizer/traversals/RangePartitionRewriter.java

```diff
@@ -0,0 +1,194 @@
+		final UnaryOperatorInformation rbOperatorInformation = new UnaryOperatorInformation(sourceOutputType, rbTypeInformation);
+		final MapPartitionOperatorBase rbOperatorBase = new MapPartitionOperatorBase(rangeBoundaryBuilder, rbOperatorInformation, "RangeBoundaryBuilder");
+		final MapPartitionNode rbNode = new MapPartitionNode(rbOperatorBase);
+		final Channel rbChannel = new Channel(sicPlanNode, TempMode.NONE);
+		rbChannel.setShipStrategy(ShipStrategyType.FORWARD, DataExchangeMode.PIPELINED);
+		final SingleInputPlanNode rbPlanNode = new SingleInputPlanNode(rbNode, "RangeBoundary PlanNode", rbChannel, DriverStrategy.MAP_PARTITION);
+		rbPlanNode.setParallelism(1);
+		rbChannel.setTarget(rbPlanNode);
+		sicPlanNode.addOutgoingChannel(rbChannel);
+		this.plan.getAllNodes().add(rbPlanNode);
+
+		// 4. Take range boundaries as broadcast input and take the tuple of partition id and record as output.
+		final AssignRangeIndex assignRangeIndex = new AssignRangeIndex(comparator);
+		final TypeInformation<Tuple2> ariOutputTypeInformation = new TupleTypeInfo<>(BasicTypeInfo.INT_TYPE_INFO, sourceOutputType);
+		final UnaryOperatorInformation ariOperatorInformation = new UnaryOperatorInformation(sourceOutputType, ariOutputTypeInformation);
+		final MapPartitionOperatorBase ariOperatorBase = new MapPartitionOperatorBase(assignRangeIndex, ariOperatorInformation, "Assign Range Index");
+		final MapPartitionNode ariNode = new MapPartitionNode(ariOperatorBase);
+		final Channel ariChannel = new Channel(sourceNode, TempMode.NONE);
+		// To avoid deadlock, set the DataExchangeMode of channel between source node and this to Batch.
+		ariChannel.setShipStrategy(ShipStrategyType.FORWARD, DataExchangeMode.BATCH);
+		final SingleInputPlanNode ariPlanNode = new SingleInputPlanNode(ariNode, "AssignRangeIndex PlanNode", ariChannel, DriverStrategy.MAP_PARTITION);
+		ariPlanNode.setParallelism(sourceParallelism);
+		ariChannel.setTarget(ariPlanNode);
+		this.plan.getAllNodes().add(ariPlanNode);
+		sourceNewOutputChannels.add(ariChannel);
+
+		final NamedChannel broadcastChannel = new NamedChannel("RangeBoundaries", rbPlanNode);
+		broadcastChannel.setShipStrategy(ShipStrategyType.BROADCAST, DataExchangeMode.PIPELINED);
+		broadcastChannel.setTarget(ariPlanNode);
+		List<NamedChannel> broadcastChannels = new ArrayList<>(1);
+		broadcastChannels.add(broadcastChannel);
+		ariPlanNode.setBroadcastInputs(broadcastChannels);
+
+		// 5. Remove the partition id.
+		final Channel partChannel = new Channel(ariPlanNode, TempMode.NONE);
+		partChannel.setDataExchangeMode(DataExchangeMode.PIPELINED);
+		final FieldList keys = new FieldList(0);
+		final boolean[] sortDirection = { true };
+		partChannel.setShipStrategy(ShipStrategyType.PARTITION_RANGE, keys, sortDirection, null, DataExchangeMode.PIPELINED);
+		ariPlanNode.addOutgoingChannel(channel);
+		partChannel.setLocalStrategy(channel.getLocalStrategy(), keys, sortDirection);
```

Comment:
We cannot remove the target node and apply its local strategy on the `PartitionIDRemoveWrapper` because
1. the keys of the local strategy are changed (here keys are fixed to `{0}` and order to `{true}`.
2. the local strategy is applied before the user functions, i.e., originally it is applied on the `SourceOut` type and here it is applied on the `Tuple2<Int, SourceOut>` type.

I would like to 
- keep the `target` node and its local strategy
- connect the `PartitionIDRemoveWrapper` to the `target` node
- change the ship strategy of the channel to the target node from `PARTITION_RANGE` to `FORWARD` and the data exchange mode to `PIPELINED`.